resources/type-system/upper-lower-bounds.md

```diff
@@ -0,0 +1,390 @@
+- **BOTTOM**(`Never`) is true
+- **BOTTOM**(`X&T`) is true iff **BOTTOM**(`T`)
+- **BOTTOM**(`X extends T`) is true iff **BOTTOM**(`T`)
+- **BOTTOM**(`T`) is false otherwise
+
+The **NULL** predicate is true for things in the equivalence class of `Null`
+
+- **NULL**(`Null`) is true
+- **NULL**(`T?`) is true iff **NULL**(`T`) or **BOTTOM**(`T`)
+- **NULL**(`T*`) is true iff **NULL**(`T`) or **BOTTOM**(`T`)
+- **NULL**(`T`) is false otherwise
+
+The **MORETOP** predicate defines a total order on top and `Object` types.
+
+- **MORETOP**(`void`, `T`) = true
+- **MORETOP**(`T`, `void`) = false
+- **MORETOP**(`dynamic`, `T`) = true
+- **MORETOP**(`T`, `dynamic`) = false
+- **MORETOP**(`Object`, `T`) = true
+- **MORETOP**(`T`, `Object`) = false
+- **MORETOP**(`T?`, `S?`) = **MORETOP**(`T`, `S`)
+- **MORETOP**(`T?`, `S`) = true
+- **MORETOP**(`T`, `S?`) = false
+- **MORETOP**(`T*`, `S*`) = **MORETOP**(`T`, `S`)
+- **MORETOP**(`T*`, `S`) = true
+- **MORETOP**(`T`, `S*`) = false
+- **MORETOP**(`FutureOr<T>`, `FutureOr<S>`) = **MORETOP**(T, S)
+
+The **MOREBOTTOM** predicate defines a total order on bottom and `Null` types.
+
+- **MOREBOTTOM**(`Never`, `T`) = true
+- **MOREBOTTOM**(`T`, `Never`) = false
+- **MOREBOTTOM**(`Null`, `T`) = true
+- **MOREBOTTOM**(`T`, `Null`) = false
+- **MOREBOTTOM**(`T?`, `S?`) = **MOREBOTTOM**(`T`, `S`)
+- **MOREBOTTOM**(`T?`, `S`) = true
+- **MOREBOTTOM**(`T`, `S?`) = false
+- **MOREBOTTOM**(`T*`, `S*`) = **MOREBOTTOM**(`T`, `S`)
+- **MOREBOTTOM**(`T*`, `S`) = true
+- **MOREBOTTOM**(`T`, `S*`) = false
+- **MOREBOTTOM**(`X&T`, `X&S`) = **MOREBOTTOM**(`T`, `S`)
+- **MOREBOTTOM**(`X&T`, `S`) = true
+- **MOREBOTTOM**(`S`, `X&T`) = false
+- **MOREBOTTOM**(`X extends T`, `X extends S`) = **MOREBOTTOM**(`T`, `S`)
+
+
+## Upper bounds
+
+We define the upper bound of two types T1 and T2 to be **UP**(`T1`,`T2`) as follows.
+
+
+- **UP**(`T`, `T`) = `T`
+- **UP**(`T1`, `T2`) where **TOP**(`T1`) and **TOP**(`T2`) = 
+  - `T1` if **MORETOP**(`T1`, `T2`)
+  - `T2` otherwise
+- **UP**(`T1`, `T2`) = `T1` if **TOP**(`T1`)
+- **UP**(`T1`, `T2`) = `T2` if **TOP**(`T2`)
+
+- **UP**(`T1`, `T2`) where **BOTTOM**(`T1`) and **BOTTOM**(`T2`) = 
+  - `T1` if **MOREBOTTOM**(`T1`, `T2`)
+  - `T2` otherwise
+- **UP**(`T1`, `T2`) = `T1` if **BOTTOM**(`T1`)
+- **UP**(`T1`, `T2`) = `T2` if **BOTTOM**(`T2`)
+
+- **UP**(`T1`, `T2`) where **NULL**(`T1`) and **NULL**(`T2`) = 
+  - `T1` **MOREBOTTOM**(`T1`, `T2`)
+  - `T2` otherwise
+
+- **UP**(`T1`, `T2`) where **NULL**(`T1`) = 
+  - `T2` if  `T2` is nullable
+  - `T2?` otherwise
+
+- **UP**(`T1`, `T2`) where **NULL**(`T2`) = 
+  - `T1` if  `T1` is nullable
+  - `T1?` otherwise
+
+- **UP**(`T1`, `T2`) where **OBJECT**(`T1`) and **OBJECT**(`T2`) =
+  - `T1` if **MORETOP**(`T1`, `T2`)
+  - `T2` otherwise
+
+- **UP**(`T1`, `T2`) where **OBJECT**(`T1`) = 
+  - `T1` if `T2` is non-nullable
+  - `T1?` otherwise
+
+- **UP**(`T1`, `T2`) where **OBJECT**(`T2`) = 
+  - `T2` if `T1` is non-nullable
+  - `T2?` otherwise
```

Comment:
Explanation related to comment on `OBJECT`: For lines 118-128, we know that `not TOP(T1)` and `not TOP(T2)`, which was the reason why I suggested not to include `dynamic` and `void` in `OBJECT`.

Reply:
ack